lib/hash-copy.js

```diff
@@ -0,0 +1,38 @@
+import { readFile, writeFile } from 'node:fs/promises';
+import { createHash } from 'node:crypto';
+import { parse as esParse } from 'espree';
+
+/**
+ * @param {URL} sourceFile
+ * @param {URL} targetDirectory
+ */
+export default async function hashCopy(sourceFile, targetDirectory) {
+  const content = await readFile(sourceFile);
+  const hash = createHash('sha256')
+    .update(content)
+    .digest('hex');
+  const originalFilename = sourceFile.pathname.split('/').pop();
+  const parts = originalFilename.split('.');
+
+  parts[parts.length - 2] = `hashed-${parts[parts.length - 2]}-${hash}`;
+
+  const hashedFileName = parts.join('.');
+  const target = new URL(targetDirectory);
+  const dependencies = [];
+
+  if (!target.pathname.endsWith('/')) {
+    target.pathname += '/';
+  }
+
+  target.pathname += hashedFileName;
+
+  for (const { type, source } of esParse(content.toString(), { ecmaVersion: 2022, sourceType: 'module' }).body) {
+    if (type === 'ImportDeclaration') {
```

Comment:
Only supports static imports, but I have some support for dynamic imports with `ImportExpression` in my local branch. The latter means the tree needs to be walked (declarations must be top level, so the code here is enough for them).